src/Spanner/ValueMapper.php

```
@@ -0,0 +1,328 @@
+    }
+
+    /**
+     * Accepts an array of key/value pairs, where the key is a SQL parameter
+     * name and the value is the value interpolated by the server, and returns
+     * an array of parameters and inferred parameter types.
+     *
+     * @param array $parameters The key/value parameters.
+     * @return array An associative array containing params and paramTypes.
+     */
+    public function formatParamsForExecuteSql(array $parameters)
+    {
+        $paramTypes = [];
+
+        foreach ($parameters as $key => $value) {
+            list ($parameters[$key], $paramTypes[$key]) = $this->paramType($value);
+        }
+
+        return [
+            'params' => $parameters,
+            'paramTypes' => $paramTypes
+        ];
+    }
+
+    /**
+     * Accepts a list of values and encodes the value into a format accepted by
+     * the Spanner API.
+     *
+     * @param array $values The list of values
+     * @return array The encoded values
+     */
+    public function encodeValuesAsSimpleType(array $values)
+    {
+        $res = [];
+        foreach ($values as $value) {
+            $res[] = $this->paramType($value)[0];
+        }
+
+        return $res;
+    }
+
+    /**
+     * Accepts a list of columns (with name and type) and a row from read or
+     * executeSql and decodes each value to its corresponding PHP type.
+     *
+     * @param array $columns The list of columns.
+     * @param array $row The row data.
+     * @return array The decoded row data.
+     */
+    public function decodeValues(array $columns, array $row)
+    {
+        $cols = [];
+        $types = [];
+
+        foreach ($columns as $index => $column) {
+            $cols[] = (isset($column['name']))
+                ? $column['name']
+                : $index;
+            $types[] = $column['type'];
+        }
+
+        $res = [];
+        foreach ($row as $index => $value) {
+            $i = $cols[$index];
```

That seems reasonable. I would prefer the default to be RETURN_ASSOCIATIVE since that is the easiest data structure to work with and in most cases people will not have empty or duplicate columns so it should be fine. I would also prefer the method to throw an error if there are duplicate fields and RETURN_ASSOCIATIVE has been provided. Almost always this will be a bug so its better to alert users rather than silently suppress one column. How does RETURN_ASSOCIATIVE handle empty field names?